### DISCUSSION TITLE
-compose Lighten_Intensity -composite adds non-existent colors when using index images

### DISCUSSION THREAD

**Discussion by bioworldmerch** [ImageMagick/ImageMagick#5067] [Digital Image Processing]
ImageMagick version

7.1.0-25

Operating system

Mac

Operating system, version and so on

10.14.6

Description
Using the command:
`Magick sourceimage.png second.png -compose Lighten_Intensity -composite result.png`
or
`Magick sourceimage.png second.png -compose Lighten -composite result.png`

When using index images limited to the same exact colors in both images, result ended up with added colors that do not exist. Any idea how I can combine the images without getting the added colors?

**Comment by snibgo** [comment]
Please show sample inputs.

**Reply by bioworldmerch** [reply]
Source:
![source](https://user-images.githubusercontent.com/84971106/164735044-27b2eeea-f6bf-48de-afef-ad915c01fa1a.png)
Second Image:
![second](https://user-images.githubusercontent.com/84971106/164735046-15cfeca6-5f0d-444e-8c0c-20c6e8d77b7a.png)

**Reply by bioworldmerch** [reply]
Compose Darken and Darken_Intensity have the same result of adding colors that do not exist as well

**Comment by snibgo** [comment]
"Darken" and "Lighten" work channel-by-channel, so can be expected to add new colours.

For "Darken_Intensity" and "Lighten_Intensity", there seems to be a bug.

The documentation at https://legacy.imagemagick.org/Usage/compose/#lighten_intensity says that Lighten-Intensity and Darken-Intensity should, for each pixel, copy the pixel from either the source or the destination. It should add no new colours to the image.

"identify -verbose" shows the two inputs shown contain:
```
light1.png:
    49798: (52,7,78) #34074E srgb(52,7,78)
    6872: (60,212,163) #3CD4A3 srgb(60,212,163)
    49851: (254,0,87) #FE0057 srgb(254,0,87)
    86865: (254,207,0) #FECF00 srgb(254,207,0)
    15654: (254,253,254) #FEFDFE srgb(254,253,254)
light2.png:
    49561: (52,7,78) #34074E srgb(52,7,78)
    6872: (60,212,163) #3CD4A3 srgb(60,212,163)
    49672: (254,0,87) #FE0057 srgb(254,0,87)
    86745: (254,207,0) #FECF00 srgb(254,207,0)
    16190: (254,253,254) #FEFDFE srgb(254,253,254)
```
The result from lighten_intensity is:
```
magick light1.png light2.png -compose Lighten_Intensity -composite  -verbose info:

    46602: (52,7,78) #34074E srgb(52,7,78)
    6894: (60,212,163) #3CD4A3 srgb(60,212,163)
    48404: (254,0,87) #FE0057 srgb(254,0,87)
    883: (254,7,78) #FE074E srgb(254,7,78)
    86865: (254,207,0) #FECF00 srgb(254,207,0)
    1333: (254,207,78) #FECF4E srgb(254,207,78)
    1047: (254,207,87) #FECF57 srgb(254,207,87)
    17012: (254,253,254) #FEFDFE srgb(254,253,254)
```
So new colours have been added, so there is a bug.

Tested with IM v7.1.0-4 on Windows 8.1.

A developer may take a look.

**Reply by bioworldmerch** [reply]
glad I could help identify it. Thank you for the response!

**Comment by snibgo** [comment]
I think the problem is in composite.c function CompositeImage(). This loops through pixels, and through the channels of each pixel. At each channel, it updates q[i], the new pixel value. However, LightenIntensity and DarkenIntensity uses "GetPixelIntensity(image,q)" to determine which input image to copy the value from. After the first channel is updated, the value of "GetPixelIntensity(image,q)" may have changed, so the next channel might be chosen from the other input image.

**Comment by snibgo** [comment]
Incidentally, whether the input is indexed is not relevant.

A simple example of the bug, using two images of one pixel each:
```
magick xc:#128 xc:#812 -compose Lighten_Intensity -composite txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (34952,8738,34952)  #888822228888  srgb(136,34,136)
```
The output pixel is neither of the input pixels.

**Comment by dlemstra** [comment]
Thanks for helping us find the cause @snibgo. Just pushed a patch to fix this. This is now the output:

```C#
C:\Users\Dirk>magick -version
Version: ImageMagick 7.1.0-31 beta Q16-HDRI x64 065f67cbc:20220419 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Debug Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193131105)

C:\Users\Dirk>magick xc:#128 xc:#812 -compose Lighten_Intensity -composite txt:
# ImageMagick pixel enumeration: 1,1,0,65535,srgb
0,0: (34952,4369,8738)  #888811112222  srgb(136,17,34)
```
